src/object/cli_obj.c

```diff
@@ -3298,21 +3304,34 @@ merge_key(d_list_t *head, char *key, int key_size)
 		}
 	}
 
-	D_ALLOC_PTR(key_one);
-	if (key_one == NULL)
+	D_ALLOC_PTR(new_key);
+	if (new_key == NULL)
 		return -DER_NOMEM;
 
-	D_ALLOC(key_one->key.iov_buf, key_size);
-	if (key_one->key.iov_buf == NULL) {
-		D_FREE(key_one);
+	D_ALLOC(new_key->key.iov_buf, key_size);
+	if (new_key->key.iov_buf == NULL) {
+		D_FREE(new_key);
 		return -DER_NOMEM;
 	}
 
-	memcpy(key_one->key.iov_buf, key, key_size);
-	key_one->key.iov_buf_len = key_size;
-	key_one->key.iov_len = key_size;
-	D_INIT_LIST_HEAD(&key_one->key_list);
-	d_list_add_tail(&key_one->key_list, head);
+	memcpy(new_key->key.iov_buf, key, key_size);
+	new_key->key.iov_buf_len = key_size;
+	new_key->key.iov_len = key_size;
+	new_key->hkey = enum_hkey_gen(obj, &new_key->key);
+	D_INIT_LIST_HEAD(&new_key->key_list);
+
+	/* Insert the key into the sorted list */
+	d_list_for_each_entry(key_one, head, key_list) {
+		if (new_key->hkey < key_one->hkey) {
```

Comment:
This is the same order for integer keys but not for hashed keys.   Does it matter?   If it doesn't matter, why even make a distinction?  Is it just so we can avoid the murmur call?  If that's the case, rather than checking if it's an integer key, why not just check if it's <= sizeof(uint64_t) and copy zero padded bytes to the value to avoid murmur hash for smaller keys altogether?

On the server, hashed keys are 16 bytes.   For keys longer than 15 bytes, it does a murmur hash and a 4 byte string hash, leaving 4 bytes empty.   One of these empty bytes is used to mark an inline key.

Reply:
Yes, hkey is for avoiding the murmur hash call.  I think we need keep both key and key for each entry in the merge list,  then later dump the key  to user kds/sgl.

And the hkey generation is just from VOS. or I miss your question?